NFT Series 9th Installement (#1138)
near/docs: docs/3.tutorials/nfts/9-series.md

---
id: series
title: Customizing the NFT Contract
sidebar_label: Lazy Minting, Collections, and More!
---

> In this tutorial, you'll learn how to take the existing NFT contract you've been working with and modify it to meet some of the most common needs I've seen in the ecosystem. This includes:

Reviewer: ```suggestion
In this tutorial, you'll learn how to take the existing NFT contract you've been working with and modify it to meet some of the most common needs in the ecosystem. This includes:
```

- Lazy Minting NFTs
- Creating Collections
- Restricting Minting Access
- Highly Optimizing Storage 
- Hacking Enumeration Methods

## Introduction

While the current NFT contract works really well for most simple use cases, it's really meant to be a foundation for creating unique and creative use-cases. The purpose of the zero to hero tutorial is to help garner a deep understanding of the tools such that you can create your own custom contracts to meet whatever use cases you might have. 

While we may not have the instructions for baking blueberry muffins, if you understand how to create the batter for regular muffins, hopefully you'll know that by simply adding blueberries to the mix or altering the amount of salt and vanilla, you'll be able to create your own delicious blueberry muffins!

<img width="45%" src="/docs/assets/nfts/customizing-logic-meme.png" />

### Collections

> One of the biggest problems with the current NFT repo is repeated data. Currently, if you mint NFTs, you'll need to store the metadata in the contract every single time. While this is fine for NFTs that have drastically different metadata, having repeated data that is similar is screaming for an optimization. In addition, one of the most common use cases for NFTs currently is *collections*.

Reviewer: ```suggestion
One of the biggest problems with the current NFT repository is repeated data. Currently, if you mint NFTs, you'll need to store the metadata in the contract every single time. While this is fine for NFTs with drastically different metadata, having repeated similar data is screaming for optimization. In addition, one of the most common use cases for NFTs currently is *collections*.
```

> A collection is a set of NFTs that are related to each other in some way. More often that not, this means they share the same media, title, description, number of copies etc. If you wanted to make a collection of 1000 NFTs with the current contract, you'd need to store this duplicate information 1000 times when in reality, the only thing that's really changing between the NFTs in the collection is the token ID. We can fix this by introducing the idea of a series. 

Reviewer: ```suggestion
A collection is a set of NFTs that are related to each other in some way. More often than not, this means they share the same media, title, description, number of copies, etc. If you wanted to make a collection of 1000 NFTs with the current contract, you'd need to store this duplicate information 1000 times when the only thing changing between the NFTs in the collection is the token ID. We can fix this by introducing the idea of a series. 
```

### Restricted Access

Currently, the NFT contract allows anyone to mint NFTs. While this works well for some projects, the vast majority of dApps and creators want to restrict who can create NFTs on the contract. This is why you'll introduce an allowlist functionality for both series and for NFTs. You'll have two data structures customizable by the contract owner:
- Approved Minters
- Approved Creators

If you're an approved minter, you can freely mint NFTs for any given series. You cannot, however, create new series.

On the other hand, you can also be an approved creator. This allows you to define new series that NFTs can be minted from. It's important to note that if you're an approved creator, you're not automatically an approved minter as well. Each of these permissions need to be given by the owner of the contract and they can be revoked at any time.

### Lazy Minting

Lazy minting is a really powerful feature that can save users from having to spend a ton of $NEAR on potentially non profitable NFTs. To understand what it is, let's look at a common scenario: 

> Benji has created an amazing digital painting of the famous Go Team gif. He wants to sell 1000 copies of it for 1 $NEAR each. Using the traditional approach, he would have to mint each copy individually and pay for the storage himself. He would then need to pay for the storage to put 1000 copies up for sale on a marketplace contract and he would need to put each up for sale **individually** which would cost a lot in Gas. After that, people would purchase the NFTs and there would be no guarantee that all or even any would be sold. There's a real possibility that nobody buys a single piece of his artwork and Benji spent all that time and effort and money on nothing. Sad boi hours... 

Reviewer: ```suggestion
Benji has created an amazing digital painting of the famous Go Team gif. He wants to sell 1000 copies of it for 1 $NEAR each. Using the traditional approach, he would have to mint each copy individually and pay for the storage himself. He would then need to pay for the storage to put 1000 copies up for sale on a marketplace contract, and he would need to put each up for sale **individually** which would cost a lot in Gas. After that, people would purchase the NFTs, and there would be no guarantee that all or even any would be sold. There's a real possibility that nobody buys a single piece of his artwork, and Benji spent all that time and effort and money on nothing. 😢  
```

```bash
git checkout main && git pull
```

You'll notice that there's a folder in the root of the project called `nft-series`. This is where the contract code lives. If you open the `src` folder, it should look similar to the following.

```
src
├── approval.rs
├── enumeration.rs
├── events.rs
├── internal.rs
├── lib.rs
├── metadata.rs
├── nft_core.rs
├── owner.rs
├── royalty.rs
├── series.rs
```

## Differences

If you sift through the code in these files, you'll notice that most of it is the same. There are only a few differences between this contract and the current NFT contract. 

### Main Library File 

Starting with `lib.rs`, you'll notice that the contract struct has been modified to now store the following information.

```diff
pub owner_id: AccountId,
+ pub approved_minters: LookupSet<AccountId>,
+ pub approved_creators: LookupSet<AccountId>,
pub tokens_per_owner: LookupMap<AccountId, UnorderedSet<TokenId>>,
pub tokens_by_id: UnorderedMap<TokenId, Token>,
- pub token_metadata_by_id: UnorderedMap<TokenId, TokenMetadata>,
+ pub series_by_id: UnorderedMap<SeriesId, Series>,
pub metadata: LazyOption<NFTContractMetadata>,
```

> Most of the information is the same, although we've added 2 new lookup sets and we've changed the `token_metadata_by_id` to be `series_by_id`.

Reviewer 2: ```suggestion
As you can see, we've replaced `token_metadata_by_id` with `series_by_id` and added two lookup sets:
```

> - **approved_minters**: Keeps track of accounts that can call the `nft_mint` function.
> - **approved_creators**: Keeps track of accounts that can create new series.
> - **series_by_id**: Map a series ID (u64) to its Series object.

Reviewer 2: ```suggestion
- **series_by_id**: Map a series ID (u64) to its Series object.

- **approved_minters**: Keeps track of accounts that can call the `nft_mint` function.
- **approved_creators**: Keeps track of accounts that can create new series.
```

In addition, we're now keeping track of a new object called a `Series`.

```rust
pub struct Series {
    // Metadata including title, num copies etc.. that all tokens will derive from
    metadata: TokenMetadata,
    // Royalty used for all tokens in the collection
    royalty: Option<HashMap<AccountId, u32>>,
    // Set of tokens in the collection
    tokens: UnorderedSet<TokenId>,
    // What is the price of each token in this series? If this is specified, when minting,
    // Users will need to attach enough $NEAR to cover the price.
    price: Option<Balance>,
    // Owner of the collection
    owner_id: AccountId,
}
```

This object stores information that each token will inherit from. This includes:
- The metadata.
- The royalties.

> - The price.

Reviewer 2: I would add links to these for further explanation / reminders

> If a price is specified, there will be no restriction on who can mint tokens in the series. In addition, if the `copies` field is specified in the metadata, **only** that number of NFTs can be minted. If the field is omitted, an unlimited amount of tokens can be minted.

Reviewer 2: I would add a "Heads Up" `:::caution` flag here

We've also added a field `tokens` which keeps track of all the token IDs that have been minted for this series. This allows us to deal with the potential `copies` cap by checking the length of the set. It also allows us to paginate through all the tokens in the series.

### Series File

> Moving on to a new file that has been introduced, you'll notice that there's no longer the `mint.rs` file and it's been replaced by a newer `series.rs` file. This is responsible for everything related to creating series as well as minting NFTs. Starting with the logic for creating a new series, you'll see a new function `create_series`.

Reviewer 2: ```suggestion
`series.rs` is a new file that replaces the old `mint.rs` logic. This is responsible for everything related to creating a series as well as minting NFTs using the `create_series` method.
```

```rust reference
https://github.com/near-examples/nft-tutorial/blob/main/nft-series/src/series.rs#L7-L58
```

> The function takes in a series ID in the form of a `u64`, the metadata, royalties, and the price for tokens in the series. It will then create the `Series` object and insert it into the contract's `series_by_id` data structure. It's important to note that the caller must be an approved creator and they must attach enough $NEAR to cover storage costs.

Reviewer 2: I would link `u64`, `Series` object, and `series_by_id` struct so people can have easy ways to learn more about something they don't yet understand. 

As much as we want users to follow a narrative that we create.... that can often not be the case as people skip to a section they want to learn about or (more likely a case) that people land on a section from a search.

#### Minting

> Next, we'll look at is the minting function. If you remember from before, this used to take a slew of parameters:
> - Token ID
> - Metadata
> - Receiver ID
> - Perpetual Royalties 
> 
> With the new minting function, these parameters have been brought down to just two.
> - The series ID
> - The receiver ID.
> 

Reviewer 2: ```suggestion
Next, we'll look at the minting function. 

In the basic version of the NFT smart contract there are several required parameters:
- Token ID
- Metadata
- Receiver ID
- Perpetual Royalties 

With the new contract, we've reduced this to two arguments:
- The series ID
- The receiver ID.

```

> Notice how the token ID isn't required? This is because the token ID is automatically generated when minting. The ID is stored on the contract as `${series_id}:${token_id}` where the token ID is a nonce that increases each time a new token is minted in a series. This not only reduces the amount of information stored on the contract but it also acts as a way to check the specific edition number.

Reviewer 2: Could add a `:::tip` banner for this section or `:::info`

> The mint function will charge users for storage and if the series has a price, it will ensure that users attach enough $NEAR to cover the price **and** storage. At the end, it will then send all the $NEAR attached minus the storage costs to the **series owner**. It's important to note that in this case, the caller does **not** need to be an approved minter.

Reviewer 2: This section is a little confusing. Let's take a stab at re-wording this to be a little more clear.

If a price wasn't specified, the $NEAR attached will be used to pay for storage and any excess will be refunded to the caller. In addition, all the callers must be approved minters.

```rust reference
https://github.com/near-examples/nft-tutorial/blob/main/nft-series/src/series.rs#L60-L149
```

### View Functions

> Now that we've introduced the idea of series, more view functions have been added to support the potentially complex use cases that arise with this contract. With these functions, a new `JsonSeries` struct has been added due to the fact that you cannot serialize an `UnorderedSet` which is being stored in each series. 

Reviewer 2: ```suggestion
Now that you are using a `Series` structure for your NFT contract, retrieving on-chain data from these contracts will require you to create new `view` methods. 

Here we will use a new struct (`JsonSeries`) that will store the metadata for each series.
```

Reviewer 2: Lmk if line 165 is accurate. Explaining the serialization of an unordered set might be unnecessary here and more appropriate in the area where we outline the `JsonSeries` struct.

<img width="80%" src="/docs/assets/nfts/series-wallet-collectibles.png" />

Hurray! You've successfully deployed and tested the series contract! **GO TEAM!**.

## Conclusion

In this tutorial, you learned how to take the basic NFT contract and iterate on it to create a complex and custom version to meet the needs of the community. You optimized the storage, introduced the idea of collections, created a lazy minting functionality, hacked the enumeration functions to save on storage, and created an allowlist functionality.

You then built the contract and deployed it on chain. Once it was on-chain, you initialized it and created two sets of series. One was complex with a price and the other was a regular series. You lazy minted an NFT and purchased it for `1.5 $NEAR` and then added yourself as an approved minter. You then minted an NFT from the regular series and viewed them both in the NEAR wallet.

> Thank you so much for going through this journey with us! I wish you all the best and am eager to see what sorts of neat and unique use-cases you can come up with. If you have any questions, feel free to ask in our Discord or any other social media channels we have.

Reviewer: ```suggestion
Thank you so much for going through this journey with us! I wish you all the best and am eager to see what sorts of neat and unique use-cases you can come up with. If you have any questions, feel free to ask on our Discord or any other social media channels we have.
```
